common/src/chain/transaction/signature/inputsig.rs

```diff
@@ -0,0 +1,170 @@
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://spdx.org/licenses/MIT
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author(s): S. Afach
+
+mod authorize_address_spend;
+mod authorize_pubkey_spend;
+
+use std::io::BufWriter;
+
+use parity_scale_codec::{Decode, Encode};
+
+use crate::{
+    chain::{ChainConfig, Destination, Transaction},
+    primitives::H256,
+};
+
+use self::{
+    authorize_address_spend::{
+        sign_address_spending, verify_address_spending, AuthorizedAddressSpend,
+    },
+    authorize_pubkey_spend::{
+        sign_pubkey_spending, verify_public_key_spending, AuthorizedPublicKeySpend,
+    },
+};
+
+use super::{
+    sighashtype::{self, SigHashType},
+    signature_hash, TransactionSigError,
+};
+
+#[derive(Debug, Encode, Decode, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub enum InputWitness {
+    #[codec(index = 0)]
+    NoSignature(Option<Vec<u8>>),
+    #[codec(index = 1)]
+    Standard(StandardInputSignature),
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Ord, PartialOrd)]
+pub struct StandardInputSignature {
+    sighash_type: SigHashType,
+    raw_signature: Vec<u8>,
+}
+
+impl StandardInputSignature {
+    pub fn new(sighash_type: sighashtype::SigHashType, raw_signature: Vec<u8>) -> Self {
+        Self {
+            sighash_type,
+            raw_signature,
+        }
+    }
+
+    pub fn get_sighash_type(&self) -> SigHashType {
+        self.sighash_type
+    }
+
+    pub fn from_data(raw_data: &Vec<u8>) -> Result<Self, TransactionSigError> {
+        let decoded_sig = StandardInputSignature::decode(&mut raw_data.as_slice())
+            .map_err(|_| TransactionSigError::DecodingWitnessFailed)?;
+        Ok(decoded_sig)
+    }
+
+    pub fn verify_signature(
+        &self,
+        chain_config: &ChainConfig,
+        outpoint_destination: &Destination,
+        sighash: &H256,
+    ) -> Result<(), TransactionSigError> {
+        match outpoint_destination {
+            Destination::Address(addr) => {
+                let sig_components = AuthorizedAddressSpend::from_data(&self.raw_signature)?;
+                verify_address_spending(chain_config, addr, &sig_components, sighash)?
+            }
+            Destination::PublicKey(pubkey) => {
+                let sig_components = AuthorizedPublicKeySpend::from_data(&self.raw_signature)?;
+                verify_public_key_spending(pubkey, &sig_components, sighash)?
+            }
+            Destination::ScriptHash(_) => return Err(TransactionSigError::Unsupported),
+        }
```

Comment:
I think output/input pair verification could be done with something like:
```rust
match (destination, witness) {
    (Destination::Pubkey(pk), Witness::Signature(sig)) => verify_pubkey_spending(pk, sig, sighash),
    (Destination::ScriptHash(sh), Witness::Script(blah)) => verify_script(sh, blah, ....);
    /* (... etc ...) */
    (_dest, _wit) => Err(TransactionError::BadDestinationWitnessPair),
}
```
This is quite nice way of ensuring we have witness type corresponding to the destination type.

Another possibility would be to keep witness as a `Vec<u8>` and decode it on demand once the destination is known into the appropriate type.

Reply:
Btw, the second option "Another possibility would be to keep witness as a Vec<u8> and decode it on demand once the destination is known into the appropriate type" is what I'm doing here anyway.